lib/core/src/Cardano/Wallet/CoinSelection/Internal/Balance.hs

```diff
@@ -816,32 +814,38 @@ performSelection = performSelectionEmpty performSelectionNonEmpty
 --          selectionHasValidSurplus constraints (transformResult result)
 --
 performSelectionEmpty
-    :: Functor m => PerformSelection m NonEmpty u -> PerformSelection m [] u
+    :: forall m ctx. (Functor m, SelectionContext ctx)
+    => PerformSelection m NonEmpty ctx
+    -> PerformSelection m []       ctx
 performSelectionEmpty performSelectionFn constraints params =
     fmap transformResult <$>
     performSelectionFn constraints (transformParams params)
   where
     transformParams
-        :: SelectionParamsOf []       u
-        -> SelectionParamsOf NonEmpty u
-    transformParams
-        = over #extraCoinSource
-            (transform (`Coin.add` minCoin) (const id))
-        . over #outputsToCover
-            (transform (const (dummyOutput :| [])) (const . id))
+        :: SelectionParamsOf []       ctx
+        -> SelectionParamsOf NonEmpty ctx
+    transformParams p@SelectionParams {..} = p
+        { extraCoinSource =
+            transform (`Coin.add` minCoin) (const id) extraCoinSource
+        , outputsToCover =
+            transform (const (dummyOutput :| [])) (const . id) outputsToCover
+        }
 
-    transformResult :: SelectionResultOf NonEmpty u -> SelectionResultOf [] u
     transformResult
-        = over #extraCoinSource
-            (transform (`Coin.difference` minCoin) (const id))
-        . over #outputsCovered
-            (transform (const []) (const . F.toList))
+        :: SelectionResultOf NonEmpty ctx
+        -> SelectionResultOf []       ctx
+    transformResult r@SelectionResult {..} = r
+        { extraCoinSource =
+            transform (`Coin.difference` minCoin) (const id) extraCoinSource
+        , outputsCovered =
+            transform (const []) (const . F.toList) outputsCovered
+        }
 
-    transform :: a -> (NonEmpty (Address, TokenBundle) -> a) -> a
+    transform :: a -> (NonEmpty (Address ctx, TokenBundle) -> a) -> a
     transform x y = maybe x y $ NE.nonEmpty $ view #outputsToCover params
 
-    dummyOutput :: (Address, TokenBundle)
-    dummyOutput = (Address "", TokenBundle.fromCoin minCoin)
+    dummyOutput :: (Address ctx, TokenBundle)
+    dummyOutput = (dummy, TokenBundle.fromCoin minCoin)
```

Comment:
Since we no longer have access to the concrete type of address, we can't create address values directly. So we rely on the `Dummy` constraint here.